lib/queuePopulator/KafkaLogConsumer/LogConsumer.js

```diff
@@ -0,0 +1,166 @@
+     * @param {string} kafkaConfig.hosts kafka hosts
+     * @param {string} kafkaConfig.topic kafka oplog topic
+     * @param {string} kafkaConfig.groupId consumer group id
+     * @param {Logger} logger logger
+     */
+    constructor(kafkaConfig, logger) {
+        const { hosts, topic, groupId } = kafkaConfig;
+        this._kafkaHosts = hosts;
+        this._topic = topic;
+        this._groupId = groupId;
+        this._log = logger;
+        this._offsets = {};
+    }
+
+    /**
+     * Get partition offsets
+     * @returns {string} stored partition offsets
+     */
+    _getOffset() {
+        // Format: [{ topic: 'oplog-topic', partition: 0, offset: 0 }]
+        return JSON.stringify(this._offsets);
+    }
+
+    /**
+     * Connects consumer to kafka and subscribes
+     * to oplog topic
+     * @param {Function} done callback
+     * @returns {undefined}
+     */
+    setup(done) {
+        // partition offsets will be managed by kafka
+        const consumerParams = {
+            'enable.auto.offset.store': true,
+            'metadata.broker.list': this._kafkaHosts,
+            'group.id': this._groupId,
+        };
+        const topicParams = {
+            'auto.offset.reset': 'earliest',
+            'enable.auto.commit': true,
+        };
+        this._consumer = new kafka.KafkaConsumer(consumerParams, topicParams);
+        this._consumer.connect();
+        this._consumer.on('ready', () => {
+            this._consumer.subscribe([this._topic]);
+            return done();
+        });
+    }
+
+    /**
+     * Waits for consumer group to rebalance
+     * @param {number} wait amount of time to wait
+     * @param {Function} cb callback
+     * @returns {undefined}
+     */
+    _waitForAssignment(wait, cb) {
+        setTimeout(() => {
+            /**
+             * assignements contain the partitions
+             * assigned to this consumer, they are only
+             * set once consumer group is balanced
+             */
+            const assignments = this._consumer.assignments();
+            if (assignments.length === 0) {
+                if (wait > maxDelayRebalance) {
+                    return cb(true);
+                }
+                return this._waitForAssignment(wait + 2000, cb);
+            }
+            return cb();
+        }, 2000);
+    }
+
+    /**
+     * Queries last commited partition offsets
+     * and stores them
+     * @param {Function} cb callback
+     * @returns {undefined}
+     */
+    _storeCurrentOffsets(cb) {
+        this._consumer.committed(5000, (err, topicPartitions) => {
+            if (err) {
+                this._log.error('Error while getting offsets', {
+                    method: 'LogConsumer._storeCurrentOffsets',
+                    topic: this._topic,
+                    groupId: this._groupId,
+                });
+                return cb(err);
+            }
+            // saving offsets
+            this._offsets = topicPartitions;
```

Comment:
why do we need to store the offset? (and where does it get stored / how to ensure we get back on track if we crash/lose it)

storing/restoring the offset should be handled already by kafka consumer api, as i understand it?

Reply:
The offsets are stored and handled entirely by Kafka, we don't manage them by hand

This was done to be compatible with how the logReader works as it requires getting the last offsets from the logConsumer and stores it when entries are published

https://github.com/scality/backbeat/blob/020213d16178fb9d2a1ae0fb41763d7bc3a7f7ea/lib/queuePopulator/LogReader.js#L553-L554